services/vod-to-live/service/src/domains/smil/generator/playlist-smil-generator.ts

```diff
@@ -0,0 +1,395 @@
+    // come back 'in network' parallel should be added for proper looping in USP
+    const firstParallel = parallels[0];
+    const lastParallel = parallels.slice(-1)[0];
+    if (
+      lastParallel?.EventStream?.Event?.Signal?.SpliceInfoSection?.SpliceInsert?.slice(
+        -1,
+      )[0]?.['@outOfNetworkIndicator'] === 1 &&
+      firstParallel?.EventStream?.Event?.Signal?.SpliceInfoSection?.SpliceInsert?.slice(
+        -1,
+      )[0]?.['@outOfNetworkIndicator'] === 1
+    ) {
+      parallels.push(
+        createParallel(
+          { video: [], audio: [] },
+          this.createEventStreamInPlaylist(0),
+        ),
+      );
+    }
+
+    return createSmilEnvelope(parallels, this.populateHeader(originalEvent));
+  }
+  private populateHeader(
+    originalEvent: PlaylistPublishedEvent,
+  ): HeaderMetadata[] {
+    const headers = [
+      ...getDefaultMetadataHeaders(new Date(originalEvent.start_date_time)),
+      createHeaderMetadata(
+        HeaderMetadataNames.MosaicPlaylistId,
+        originalEvent.id,
+      ),
+    ];
+    if (this.drmSettings.decryptionCpixFile) {
+      headers.push(
+        createHeaderMetadata(
+          HeaderMetadataNames.DecryptCpix,
+          this.drmSettings.decryptionCpixFile,
+        ),
+      );
+    }
+    if (this.drmSettings.encryptionDashCpixFile) {
+      headers.push(
+        createHeaderMetadata(
+          HeaderMetadataNames.MpdCpix,
+          this.drmSettings.encryptionDashCpixFile,
+        ),
+      );
+    }
+    if (this.drmSettings.encryptionHlsCpixFile) {
+      headers.push(
+        createHeaderMetadata(
+          HeaderMetadataNames.HlsCpix,
+          this.drmSettings.encryptionHlsCpixFile,
+        ),
+      );
+    }
+    return headers;
+  }
+
+  /**
+   * Transform playlist program into SMIL parallels with the insertions of ads.
+   * @param program Playlist program (VOD video).
+   * @returns - SMIL parallels in order they should be played in.
+   */
+  private processProgram(program: Program): Parallel[] {
+    const parallels: Parallel[] = [];
+    // getting audio and video streams of the video
+    const programReference = videoToSmilParallelReferences(
+      program.video,
+      this.mutualStreams,
+    );
+
+    const preCuePoint = program.program_cue_points?.filter(
+      (cp) => cp.type === 'PRE',
+    )[0];
+    const midCuePoints =
+      program.program_cue_points
+        ?.filter((cp) => cp.type === 'MID')
+        .sort((cp1, cp2) => {
+          return (
+            (cp1.time_in_seconds != null ? cp1.time_in_seconds : 0) -
+            (cp2.time_in_seconds != null ? cp2.time_in_seconds : 0)
+          );
+        }) ?? [];
+    const postCuePoint = program.program_cue_points?.filter(
+      (cp) => cp.type === 'POST',
+    )[0];
+
+    // adding PRE cue point
+    if (preCuePoint) {
+      parallels.push(...this.processCuePoint(preCuePoint));
+    }
+
+    //cutting program video for MID cue points
```

Comment:
I understand the importance of maintaining consistency in the code, but I would like to request that this particular change be handled separately. The current task already involves modifying a large number of files, and adding this change on top of it may result in a less clean or organized pull request.

Reply:
Separate work item 40235 created